modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,346 @@
+
+   Copyright © 2006 Ruslan Savchenko <savrus@mexmat.net>
+   Copyright © 2010 Shane Tzen <shane@ict.usc.edu>
+   Copyright © 2012-2020 Sven Hartge <sven@svenhartge.de>
+   Copyright © 2016 Keller Fuchs <kellerfuchs@hashbang.sh>
+
+*/
+
+#include <sys/types.h>
+#include <sys/quota.h>
+#include <linux/quota.h>
+#include <pwd.h>
+#include <syslog.h>
+#include <ctype.h>
+#include <errno.h>
+#include <mntent.h>
+#include <stdio.h>
+
+#define PAM_SM_SESSION
+
+#include <security/pam_modules.h>
+#include <security/_pam_macros.h>
+#include <security/pam_ext.h>
+#include <security/pam_modutil.h>
+#include "pam_inline.h"
+
+#ifndef PATH_LOGIN_DEFS
+# define PATH_LOGIN_DEFS "/etc/login.defs"
+#endif
+
+#define MAX_UID_VALUE 0xFFFFFFFFUL
+
+#ifndef TRUE
+# define TRUE 1
+#endif
+#ifndef FALSE
+# define FALSE 0
+#endif
+
+struct pam_params {
+  uid_t start_uid;
+  uid_t end_uid;
+  const char *fs;
+  size_t fs_len;
+  int overwrite;
+  int debug;
+};
+
+static inline void debug(pam_handle_t *pamh, struct if_dqblk *p, char *device) {
+  pam_syslog(pamh, LOG_DEBUG, "device=%s bsoftlimit=%llu bhardlimit=%llu "
+                              "isoftlimit=%llu ihardlimit=%llu",
+             device, p->dqb_bsoftlimit, p->dqb_bhardlimit, p->dqb_isoftlimit,
+             p->dqb_ihardlimit);
+}
+
+static int _pam_parse_dqblk(int argc, const char **argv, struct if_dqblk *p) {
+  int bhard = FALSE, bsoft = FALSE, ihard = FALSE, isoft = FALSE;
+
+  /* step through arguments */
+  for (; argc-- > 0; ++argv) {
+    const char *str;
+    char *ep = NULL;
+    if ((str = pam_str_skip_prefix(*argv, "bhardlimit=")) != NULL) {
+      p->dqb_bhardlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_bhardlimit < 0))
+        p->dqb_bhardlimit = 0;
+      p->dqb_valid |= QIF_BLIMITS;
+      bhard = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "bsoftlimit=")) != NULL) {
+      p->dqb_bsoftlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_bsoftlimit < 0))
+        p->dqb_bsoftlimit = 0;
+      p->dqb_valid |= QIF_BLIMITS;
+      bsoft = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "ihardlimit=")) != NULL) {
+      p->dqb_ihardlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_ihardlimit < 0))
+        p->dqb_ihardlimit = 0;
+      p->dqb_valid |= QIF_ILIMITS;
+      ihard = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "isoftlimit=")) != NULL) {
+      p->dqb_isoftlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_isoftlimit < 0))
+        p->dqb_isoftlimit = 0;
+      p->dqb_valid |= QIF_ILIMITS;
+      isoft = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "btime=")) != NULL) {
+      p->dqb_btime = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_btime < 0))
+        p->dqb_btime = 0;
+      p->dqb_valid |= QIF_BTIME;
+    } else if ((str = pam_str_skip_prefix(*argv, "itime=")) != NULL) {
+      p->dqb_itime = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_itime < 0))
+        p->dqb_itime = 0;
```

Comment:
All these members of ```struct if_dqblk``` have an unsigned 64-bit type, so they always >= 0.
You probably want to use ```strtoll``` instead of ```strtol``` and assign its result to a temporary variable of type ```long long```.
I also suggest to create a helper function taking a ```const char *``` argument and returning an ```unsigned long long``` result, this would make the parser shorter and simpler.

Reply:
Module internal helper functions names should start with "_", correct?

Reply:
> Module internal helper functions names should start with "_", correct?

No, but they have to be static.

Reply:
So:
```
static unsigned long long
_pam_setquota_convert_num(const char *str) {
  long long temp;
  char *ep = NULL;

  temp = strtoll(str, &ep, 10);
  if (!ep || (temp < 0)) 
    return 0
  else
    return temp;
}
```
`return` should to an automatic cast, if I am not wrong and since because of the preceding `if` `temp` has to be >= 0, this is safe.

Reply:
> No, but they have to be static.
Yes, of course. But since `_pam_parse_dqblk` already exists, I am keeping the naming somewhat consistent.

Reply:
No leading underscores, please, this is a reserved namespace.
You don't need pam_setquota prefix either.
convert_num looks too generic, maybe some mentioning of dqb won't harm.
I'd suggest something in lines of ```str_to_dqb_num```.
You can just write
```
long long num = strtoll(str, &ep, 10);
```
P.S. there has to be a semicolon after ```return 0```.

Reply:
I will rename the other functions then as well.

Reply:
> > No, but they have to be static.
> 
> Yes, of course. But since `_pam_parse_dqblk` and `_pam_parse_params` already exist, I am keeping the naming somewhat consistent.

They shouldn't exist either, please rename them, simple ```parse_dqblk``` and ```parse_params``` would do.

Reply:
Done.